tee/tee.go

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package tee
+
+import (
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+type (
+	Enclave interface {
+		// GenKeys generates a new secp256k1 ECDSA key and stores it as the
+		// enclave's signing key.
+		//
+		// It returns the public key derived Ethereum address and attestation of
+		// correct initialization of the enclave with the generated address. The
+		// attestation can be used to verify the enclave with the TEE vendor.
+		//
+		// The Operator must deploy the contract with the Enclave's address after
+		// calling GenKeys.
+		GenKeys() (common.Address, []byte, error)
+
+		// ProcessBlocks should be called by the Operator to cause the enclave to
+		// process the given block(s), logging deposits and exits.
+		//
+		// Note that BalanceProofs requires an additional k blocks to be known to
+		// the Enclave before it reveals an epoch's balance proofs to the operator.
+		// k is a security parameter to guarantee enough PoW depth.
+		ProcessBlocks(...Block) error
+
+		// ProcessTXs should be called by the Operator whenever they receive new
+		// transactions from users. After a transaction epoch has finished and an
+		// additional k blocks made known to the Enclave, the epoch's balance proofs
+		// can be received by calling BalanceProofs.
+		ProcessTXs(...Transaction) error
+
+		// BalanceProofs returns the balance proofs of the given epoch. Note that
+		// all blocks of the epoch's transaction phase (+k) need to be known to the
+		// Enclave.
+		BalanceProofs(Epoch) ([]Proof, error)
+	}
+
+	// Epoch is the epoch counter type.
+	Epoch = uint64
+
+	// A Block is a go-ethereum block together with its receipts. go-ethereum's
+	// types.Block type doesn't store the receipts...
+	Block struct {
+		types.Block
+		Receipts types.Receipts
+	}
+
+	Transaction struct {
```

Comment:
Doesn't this need a sender and recipient?

Reply:
and what about user signatures?

Reply:
You're totally right :joy: